src/lib/when.ts

```diff
@@ -0,0 +1,92 @@
+/**
+ * @license
+ * Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {directive, reparentNodes, DirectiveFn, NodePart} from '../core.js';
+
+type PartCache = [NodePart, NodePart, DocumentFragment];
+const partCaches = new WeakMap<NodePart, PartCache>();
+
+/**
+ * Efficiently switches between two templates based on the given condition. The rendered
+ * content is cached, and re-used when switching conditions. Templates are evaluated
+ * lazily, so the passed values must be functions.
+ *
+ * While this directive can render any regular part, it makes the most sense when used
+ * with TemplateResulte since most other values are dirty checked already.
+ *
+ * Example:
+ *
+ * let checked = false;
+ *
+ * html`
+ *   when(checked, () => html`Checkmark is checked`, () => html`Checkmark is not checked`);
+ * `
+ *
+ * @param condition the condition to test against
+ * @param ifValue the content to render if
+ * @param elseValue
+ */
+export const when = (condition: boolean, ifValue: () => any, elseValue: () => any): DirectiveFn<NodePart> =>
+    directive((part: NodePart) => {
+      // Create cache if this was the first render
+      if (!partCaches.has(part)) {
+        createPartCache(part);
+      }
+
+      const partCache = partCaches.get(part)!;
+      const cacheContainer = partCache[2];
+      const parts = [partCache[0], partCache[1]];
+      let value;
+
+      if (condition) {
```

Comment:
Before even checking the condition, I think we want to check that it's different from last time. If it's the same we can just propagate the current value.

Reply:
The value to set depends on the condition, even if it didn't change. So we still need to do some work. You're right though currently we do a lot of dom manipulation when not necessary.

I've separated the work for moving dom nodes when switching conditions.